src/projects/shared/api/circuit/tests/public/Selections.test.ts

```diff
@@ -4,40 +4,329 @@ import "../Extensions";
+        const w2 = Connect(c1, c3);
+        const p1 = GetPort(c1);
+        const p2 = GetPort(c2);
+        const p3 = GetPort(c3);
+        c1.select();
+        c2.select();
+        p3.select();
+        w1.select();
+        test(".all", () => {
+            expect(selections.all).toHaveLength(4);
+            expect(selections.all).toContain(c1);
+            expect(selections.all).toContain(c2);
+            expect(selections.all).toContain(p3);
+            expect(selections.all).toContain(w1);
+        });
+        test(".components", () => {
+            expect(selections.components).toHaveLength(2);
+            expect(selections.components).toContain(c1);
+            expect(selections.components).toContain(c2);
+        });
+        test(".wires", () => {
+            expect(selections.wires).toHaveLength(1);
+            expect(selections.wires).toContain(w1);
+        });
     });
 
     describe("Midpoint", () => {
-        // Test ideas:
-        // Will have to do some math (don't have to do anything crazy)
+        test("2 components selected", () => {
+            const [{selections}, { }, {PlaceAt, Connect, GetPort}] = CreateTestRootCircuit();
+            const [c1, c2] = PlaceAt(V(0, 0), V(2, 2));
+            c1.select();
+            c2.select();
+            expect(selections.midpoint()).toStrictEqual(V(1, 1));
+        });
+        test("6 components selected", () => {
+            const [{selections}, { }, {PlaceAt, Connect, GetPort}] = CreateTestRootCircuit();
+            const [c1, c2, c3, c4, c5, c6] = PlaceAt(V(0, 0), V(2, 2), V(0, 1), V(2, 1), V(1, 0), V(1, 2));
+            c1.select();
+            c2.select();
+            c3.select();
+            c4.select();
+            c5.select();
+            c6.select();
+            expect(selections.midpoint()).toStrictEqual(V(1, 1));
+        });
+        test("Use bounding box midpoint, not weighted average", () => {
+            const [{selections}, { }, {PlaceAt, Connect, GetPort}] = CreateTestRootCircuit();
+            const [c1, c2, c3,] = PlaceAt(V(0, 0), V(2, 2), V(0, 0));
+            c1.select();
+            c2.select();
+            c3.select();
+            expect(selections.midpoint()).toStrictEqual(V(1, 1));
+        });
     });
 
     describe("Clear", () => {
-        // Test ideas:
-        // Make sure undo/redo works for this
+        test("Basic with undo/redo", () => {
+            const [{selections, undo, redo}, { }, {PlaceAt, Connect, GetPort}] = CreateTestRootCircuit();
+            const [c1, c2, c3] = PlaceAt(V(0, 0), V(1, 1), V(2, 2));
+            const w1 = Connect(c1, c2);
+            const w2 = Connect(c1, c3);
+            const p1 = GetPort(c1);
+            const p2 = GetPort(c2);
+            const p3 = GetPort(c3);
+            c1.select();
+            c2.select();
+            p3.select();
+            w1.select();
+            expect(selections.all).toHaveLength(4);
+            expect(selections.all).toContain(c1);
```

Comment:
You can use `.toContainObjsExact([c1, c2, p3, w1])` to make sure it contaisn exactly the set of given of objects, or `.toContainObjs` to see if it at *least* contains the given objects instead of having an `expect` for each